### ISSUE TITLE
Having over 2000+ items stored in storage systems causes chunk "corruption"

### ISSUE THREAD

**Issue by negativize** [blushiemagic/MagicStorage#41]
This issue seems to be consistent and has been reported numerous times on the thread on the Terraria forums. It also seems to be multiplayer-only, as the issued isn't present in singleplayer. The corruption appears to happen after exiting a world with 2000+ items stored.

Problem: when over 2000 or more items total are stored in a system and a server is restarted, everything in the spawn chunk will be erased upon restart _(might be chunks around storage that are affected, not sure)_

I don't believe this affects wired items (as our wired doors were not erased when the issue happened), but everything else is deleted. I believe this also happens when multiple system have a total storage of 2000+ items. This does not happen when there are over 2000 slots _available._

Strangely, NPCs inside the deleted chunk fall are not affected and instead teleport to the position of where there homes were after falling a certain distance.

This corruption doesn't actually "delete" items: in singleplayer, any affected chunks return to their former state and any storage units work as intended. The only way to fix the issue (with my knowledge) is to start the corrupted world in single-player, remove items until less than 2000 are present, and restart the affected world. This fixes the issue and everything is returned to its prior state.

Image taken from forum in singleplayer:
![image](https://user-images.githubusercontent.com/37283679/37266222-d24492d4-2586-11e8-8377-fc79b3e39c92.png)

Image taken from forum in multiplayer:
![image](https://user-images.githubusercontent.com/37283679/37266257-14f4af6a-2587-11e8-9faf-f22615964b5f.png)

Some users are reporting this happens at different limits. (one user reported 8000 - 8500 items being the limit) I would only consider this issue critical in multiplayer, as it basically ruins the gimmick of storage by limiting the server to less than 2000 items. Works fine in singleplayer. Either way, it needs a fix.

**Comment by ukilop** [comment]
same issue here, temp work around is have the storage containers in different chunks.
in my multiplayer world we moved a bit from spawn and put 4 sets of 5 (hellstone) containers with each set being in a different chunk, fyi the problem doesn't exist for single player so simple take the multiplayer world and open it in single player and move some storage around, then launch host&play to test if it worked

**Comment by blushiemagic** [comment]
bcd349952a435274d3a85fbb42dbad977a482975 should delay this from happening. However I'll still leave this issue open since it's only a temporary fix, and it's planned for the next tModLoader update to fix this permanently.

**Comment by Tenrys** [comment]
This still happens today, having mods like "Even more modifiers" and "Spirit Mod" (with glyphs) makes this happen even quicker. Not sure what the course of action is, I hardly doubt "take the items out in single player and split them through multiple chunks" is a viable solution at all.

**Comment by Tenrys** [comment]
I tried to host another world with less mods and a tiny amount of MagicStorage containers, and the same deal happened. It only seems to happen on my Linux Dedicated Server, when my friends try to host it it just works without an issue.

**Comment by Psyda** [comment]
![Linux VPS with 2 MS systems totaling 450-600 items](https://user-images.githubusercontent.com/17040097/47480227-b8af6400-d7fd-11e8-83da-b0c4e8582cb3.png)

![Windows server with same mods and 12,000+ items.](https://user-images.githubusercontent.com/17040097/47480223-b64d0a00-d7fd-11e8-98ee-d0dd03203d7e.png)

I think the issue is deeper than expected.

**Comment by MarkusGordathian** [comment]
Where exactly is `libMonoPosixHelper.so` found? Does it come with tModLoader or do I have to download it myself?

**Comment by Silly511** [comment]
You need to extract them from a mono install. [Here's](https://github.com/blushiemagic/MagicStorage/files/3001084/Libs.zip) a zip with a mac and linux lib. I'm not sure if the linux lib is 64 bit or 32 bit, I just installed mono in a linux virtual machine and grabbed it from the libs directory.

**Comment by MarkusGordathian** [comment]
It looks like those libraries are 64-bit. I tried both those and copying the library from my own mono installation, but to no avail. The chunk rendered, but the storage system itself still doesn't work. I can deposit items, but it displays "0/0 items" and I can't retrieve them. However, I can retrieve items in singleplayer, which is very confusing.

**Comment by drax32** [comment]
I have this problem as well on linux as a client, how do I fix this?

**Comment by ultrako** [comment]
Yep, those files are 64 bit. I put the library file in Terraria's lib64 folder, and can successfully connect to a server with magic storage. I'm on a linux client, and can access all of the items in the Storage Heart.

**Comment by halcyonesselion** [comment]
Would anyone be able to provide a download for a 32-bit `libMonoPosixHelper.dylib`? I'm using tModLoader to host these mods and am having the same issue, but the lib in Silly511's zip folder didn't solve the problem. I suspect it's because the current stable version of tModLoader is 32-bit. All other issues are the same.

**Comment by CobaltWarrior9** [comment]
Hello, I would like to add to this problem by saying I am having this same problem on Windows. I have also had this problem after placing 3 empty storage units with no heart side by side. After about 5-10 minutes, my game crashed, I opened it back up (on multiplayer) and the chunk was gone.

**Comment by Steffo99** [comment]
I'd suggest pinning this issue, since it seems to be a common problem.

**Comment by Steffo99** [comment]
@Silly511 Thanks, your fix worked on Ubuntu 18.04LTS with tModLoader v0.11.3!

**Comment by syy1125** [comment]
Encountered the world hole issue, @Silly511 thanks for you fix - it worked for me.
I'm going to look through the open issues and point others having similar problems to this thread.

**Comment by syy1125** [comment]
Also, https://github.com/blushiemagic/MagicStorage/issues/74#issuecomment-484970403 mentioned that v0.4.3.1 was working for Linux but v0.4.3.2 and later were not. I found commit https://github.com/blushiemagic/MagicStorage/commit/bcd349952a435274d3a85fbb42dbad977a482975 in the history of v0.4.3.2, so I'd say https://github.com/blushiemagic/MagicStorage/issues/41#issuecomment-450196167 is on point.

**Comment by ivakam** [comment]
I'm getting wildly varying results with the fix mentioned earlier in the thread. At first it seemed to be working, but now I'm either getting stuck on "Receiving tile data: Complete!" or the client connects, but then freezes and crashes. The steps I've taken is copying the .so file to both lib and lib64.

For reference, I'm on Fedora 30 with modloader 0.11.3 and Magic Storage 0.4.3.4.

**Comment by warren-fisher** [comment]
I am using Debian with tmodloader 0.11.5 and it is not working. Similar problem to @ivakam with the .so fix, I get stuck trying to connect at 33% or 40%. I will look into MagicStorage and see if I can find a fix.

**Comment by Reutertu3** [comment]
I can attest that placing libMonoPosixHelper.so into the lib64 library solved the issue for me (Debian 10). Despite using the latest tmodloader (0.11.6.2) that particular library wasn't present which evidently caused the issue.

**Comment by MCwindTIM** [comment]
I got this issue too, Is anyone knows how to fix this on Windows 10 1903?
Magic Storage v0.4.3.5 with tModLoader v0.11.7.1
![image](https://user-images.githubusercontent.com/26113926/82600704-ec708c80-9be0-11ea-975c-8dd657d45485.png)

**Comment by tzhouhc** [comment]
tModloaderServer 0.11.7.5, magic storage 0.4.3.5, same issue as described above, system is ubuntu 18.04, can confirm that `libMonoPosixHelper.so` was not present in tModLoaderServer's linux release and the issue was fixed after manually including the lib file.

**Comment by TNTftw21** [comment]
Just had this issue rear its ugly head again. tModLoader v0.11.8.4, Magic Storage 0.4.3.5. I can't seem to find where tModLoader logs full errors (instead I only get the "Error on message Terraria.MessageBuffer" errors), but I believe this is the same problem. Strangely, there were only a handful of items - somewhere in the neighborhood of 250 items including the chests I hadn't merged into the system yet. Relogging without restarting the server (I run the server in a terminal to make finding and testing things like this easier) caused a chunk of the world to disappear for me, but my two friends running Windows were just fine. Restarting the server and relogging caused all of the blocks to no longer follow visual attachment rules, and getting close to where the chunk of the world went missing caused the client to crash. Unlike previous times I've encountered this problem, tML 0.11.8.4 _does_ have the appropriate 32-bit and 64-bit versions of libMonoPosixHelper.so (I also made sure it was using the provided versions and not my native ones using `pmap -p`). Swapping in the file provided by Fadenfire did not fix the bug. Everything works fine in singleplayer.

**Comment by hollowharkov** [comment]
Same issue on Ubuntu 20.04, tmodloader 0.11.8.4, Magic Storage 0.4.3.5. TIles aren't connecting properly and chunks are missing around the storage system. Items are also missing from the system until I restart my game. I also tried default libMonoPosixHelper.so and the provided version to no avail. Works fine in singleplayer for me as well.

**Comment by TNTftw21** [comment]
[libMonoPosixHelper.zip](https://github.com/blushiemagic/MagicStorage/files/6797815/libMonoPosixHelper.zip)
This is the version of the library I'm using. It's provided by mono-runtime-common version 6.12.0.122-0xamarin1+ubuntu2004b1. I literally just copied it from `/lib` to the tML `lib64` directory.

**Comment by hollowharkov** [comment]
I can confirm that the fix from TNTftw21 works for me as well, hopefully tmodloader uses that lib in a future update.

**Comment by mazunki** [comment]
> [libMonoPosixHelper.zip](https://github.com/blushiemagic/MagicStorage/files/6797815/libMonoPosixHelper.zip) This is the version of the library I'm using. It's provided by mono-runtime-common version 6.12.0.122-0xamarin1+ubuntu2004b1. I literally just copied it from `/lib` to the tML `lib64` directory. It got me through the rest of that playthrough without issue.

Both of the LD_PATHS under `$STEAM_PATH/tModLoader` included the `libMonoPosixHelper.so` for me, the x86 version being 263K and the 64-bit being 261K in size. Should I override the 64-bit with the older version? I didn't even have mono installed until recently. How was my game even working?

My issue is not about chunk corruption, per se, but rather inability to find certain items in the storage. Does anyone know if the cause the same?

**Comment by flarn2006** [comment]
Does this happen in Crimson worlds too?

**Comment by mazunki** [comment]
> Does this happen in Crimson worlds too?

My worldgen has Fargowilta's Both Evils installed, but if I recall correctly the world "is" a corrupted one. I would be surprised if that's related at all, @flarn2006 .

**Comment by aMytho** [comment]
I'm also having the `Error on message Terraria.MessageBuffer`. I am using terraria 1.3.x with the 64bit version of tmodloader. (Linux server). The complete version of mono is installed. I've ran the exe and the script but both methods have the same problem.

I check my linux `lib` folder and I see the `libMonoPosixHelper.so` file. However, I don't seem to have a lib64 folder. I created one in the root directory and added the file to it but that doesn't seem to change anything. I've only tried it with a few items and it seems to work but the error still appears in the logs.

**Comment by SpoopyTim** [comment]
> Both of the LD_PATHS under `$STEAM_PATH/tModLoader` included the `libMonoPosixHelper.so` for me, the x86 version being 263K and the 64-bit being 261K in size. Should I override the 64-bit with the older version? I didn't even have mono installed until recently. How was my game even working?
> 
> My issue is not about chunk corruption, per se, but rather inability to find certain items in the storage. Does anyone know if the cause the same?

Can confirm the same as mazunki

Not having any chunk issues but intense lag when opening storage and even more lag when opening the crafting interface.

There's also the disappearance of items at random, some items that should have been in there from a long time ago are gone and some items that i just put in disappear too. Both cases of which some are permanent and some come back after a server restart.

~5000 items in storage

Running a docker container on Ubuntu 22.04 running on a high spec machine
[Dockerfile](https://github.com/tModLoader/tModLoader/blob/1.4.3-dev-archive/patches/tModLoader/Terraria/release_extras/DedicatedServerUtils/Docker/Dockerfile)
`docker-compose.prod.yml`
```docker-compose
version: "1.0"
services:
  tml:
    # The built docker image
    image: tmodloader-server_tml
    # Allow attach
    stdin_open: true
    tty: true
    ports:
      - 7189:7777
    volumes:
      - "./TModLoader-Server/Terraria:/home/tml/.local/share/Terraria"
```

**Comment by rafapousin** [comment]
![image](https://github.com/blushiemagic/MagicStorage/assets/37030637/ee2f1f6a-cd94-4aaa-b7c0-31c809585d75)
The chunks missing are the grey are that goes around in a square shape, u can see how it goes down passing the hell (completelly removed)

It just happend to me today, kinda worried tho in singleplayer its completelly fine, tried to see if it was some time of incompatibility with any biome generating mod, but the bug ate an insane ammount of chunks, mushroom biome and hell included, i even fall to the void and died.

![image](https://github.com/blushiemagic/MagicStorage/assets/37030637/91a44007-9677-4ad7-9a30-487543349268)
(list of the mods installed along with magic storage)

**Comment by ExterminatorX99** [comment]
Considering how MUCH is missing I feel this isn't caused by magic storage. Could you send your server log and client log?

**Comment by rafapousin** [comment]
> Considering how MUCH is missing I feel this isn't caused by magic storage. Could you send your server log and client log?

yeah, ive been thinking that too since the storage chunk is not missing and i dont have 2000+ stacks of items stored, maybe i was a bit fast with the post, after restartign the server the chunks reapeared again, same with opening the world in singleplayer (in case the logs found as "client" and "server" in setamapps/commom/tModLoader/tModLoader-Logs are worthless)
[server.log](https://github.com/blushiemagic/MagicStorage/files/13059131/server.log)
[client.log](https://github.com/blushiemagic/MagicStorage/files/13059213/client.log)

Im gonna guess with my null knowledge about terraria logs that those that ive adjunted do not containt he error since as i mentioned early, i restarted the server and it worked good, checking the "Old" logs i found this one 
[server.log](https://github.com/blushiemagic/MagicStorage/files/13059207/server.log) that im gonna asume is the log made after the "bug" since its quite larger than the other "old" logs.

**Comment by ExterminatorX99** [comment]
Yeah, that second log is the bugged one.
I see a bunch of Calamity exceptions in there.
Best report it to them.
Glad to hear a restart fixed the issue.

**Comment by rafapousin** [comment]
will do, ty

**Comment by absoluteAquarian** [comment]
The following message has been copy/pasted to all issues older than 3 months.

In order to clean up how the project looks, I'll be closing all old issues.  If you believe that this issue is still relevant, please make a new issue that references this issue in its message.
